### PR TITLE
Split question library to break circular dependency

### LINKED ISSUE #8875: Break circular dependency between external grading socket and question library
This is a circular dependency:

```
warn no-circular: apps/prairielearn/src/lib/externalGrader.ts → 
      apps/prairielearn/src/lib/externalGradingSocket.js →
      apps/prairielearn/src/lib/question.js →
      apps/prairielearn/src/lib/externalGrader.ts
```

This is because externalGrader.ts sends a job status update to externalGradingSocket; the socket uses question.js to render the panels after grading is complete; and question.js uses externalGrader.ts to trigger the start of a new grading process. I believe the best approach to break this cycle is to break down question.js, particularly the grading functions from the rendering functions.

_Originally posted by @jonatanschroeder in https://github.com/PrairieLearn/PrairieLearn/issues/8872#issuecomment-1821071588_

### PR DESCRIPTION
**Pull request by jonatanschroeder**
While the original reason to do this was to resolve #8875, a bonus fact is that the code is now easier to navigate. In essence, the question.js library was split into ~~three~~ four:

* grading.js includes `saveSubmission`, `saveAndGradeSubmission`, and `gradeVariant`.
* question-testing.js includes `startTestQuestion` and its helper functions.
* question-render.js includes `getAndRenderVariant`, `renderPanelsForSubmission`, `setRendererHeader` and related helpers.
* question-variant.js includes `getFile`, `ensureVariant` and `getQuestionCourse`.
* `writeCourseIssues` is moved to issues.js.

The functions were not modernized yet, to simplify the process.

### CODE REVIEW

File: apps/prairielearn/src/lib/externalGradingSocket.js [hunk 1]
```diff
@@ -127,34 +129,6 @@ module.exports.gradingJobStatusUpdated = function (grading_job_id) {
...
-    csrfToken,
-    authorizedEdit,
-    true, // renderScorePanels
-    (err, results) => {
-      if (ERR(err, callback)) return;
-      callback(null, results);
-    },
-  );
-};
-
```
Comment (jonatanschroeder) [inline apps/prairielearn/src/lib/externalGradingSocket.js:157]: This function was not used anywhere else in the code, so I changed it's only call to go directly to the lib.

**Review by nwalters512 — COMMENTED** [review]
I don't love that this results in losing the Git history of some of our most foundational code. Have you considered a method like the one described here to retain history even after splitting files? https://stackoverflow.com/questions/3887736/keep-git-history-when-splitting-a-file

I won't necessarily block on this, but I would like to explore our options. I use `git blame` via my editor many times a day and I really value being able to quickly see when a line was changed and why.

**Comment by jonatanschroeder** [comment]
> I don't love that this results in losing the Git history of some of our most foundational code. Have you considered a method like the one described here to retain history even after splitting files? https://stackoverflow.com/questions/3887736/keep-git-history-when-splitting-a-file
> 
> I won't necessarily block on this, but I would like to explore our options. I use `git blame` via my editor many times a day and I really value being able to quickly see when a line was changed and why.

@nwalters512 The commits have been rearranged to allow a regular merge or rebase merge to keep history. Locally this tracks history in my tests. If you'd prefer separate PRs let me know, though the intermediate commit does not pass the tests since it's a simple rename.

**Review by nwalters512 — APPROVED** [review]
Note that this *must* not be merged with squashing. I'm going to temporarily enable "rebase and merge" and admin-override to perform the merge that way.

**Pull request merged**